Sources/DistributedSystem/ChannelCompression.swift

```diff
@@ -0,0 +1,202 @@
+            self.timer = nil
+        }
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        if let timer {
+            timer.cancel()
+            self.timer = nil
+        }
+        var buffer = unwrapInboundIn(data)
+        if let compression = buffer.readInteger(as: UInt8.self) {
+            if compression != 0 {
+                _ = context.pipeline.addHandler(ChannelCompressionInboundHandler(loggerBox), position: .after(streamHandler))
+                _ = context.pipeline.addHandler(ChannelCompressionOutboundHandler(), position: .last)
+            }
+            context.fireChannelActive()
+            if buffer.readableBytes > 0 {
+                context.fireChannelRead(wrapInboundOut(buffer))
+            }
+            _ = context.pipeline.removeHandler(self)
+        } else {
+            let channel = context.channel
+            logger.info("Invalid compression request received from client @ \(channel.remoteAddressDescription), close connection.")
+            channel.close(promise: nil)
+        }
+    }
+}
+
+final class ChannelCompressionHandshakeClient: ChannelInboundHandler, RemovableChannelHandler {
+    typealias InboundIn = ByteBuffer
+
+    private let loggerBox: Box<Logger>
+    private let compression: UInt8
+    private let streamHandler: NIOCore.ChannelHandler
+
+    init(_ loggerBox: Box<Logger>, _ compression: UInt8, _ streamHandler: NIOCore.ChannelHandler) {
+        self.loggerBox = loggerBox
+        self.compression = compression
+        self.streamHandler = streamHandler
+    }
+
+    func channelActive(context: ChannelHandlerContext) {
+        var buffer = ByteBufferAllocator().buffer(capacity: MemoryLayout<UInt8>.size)
+        buffer.writeInteger(compression, as: UInt8.self)
+        context.writeAndFlush(NIOAny(buffer), promise: nil)
+        if compression != 0 {
+            _ = context.pipeline.addHandler(ChannelCompressionInboundHandler(loggerBox), position: .after(streamHandler))
+            _ = context.pipeline.addHandler(ChannelCompressionOutboundHandler(), position: .last)
+        }
+        context.fireChannelActive()
+        _ = context.pipeline.removeHandler(self)
+    }
+}
+
+fileprivate struct BufferManager {
+    var buffers = Array(repeating: UnsafeMutableRawBufferPointer(start: nil, count: 0), count: 2)
+    var nextIdx = 0
+
+    mutating func getNextBuffer(capacity: Int) -> UnsafeMutableRawBufferPointer {
+        let idx = nextIdx
+        nextIdx = (1 - nextIdx)
+        var buffer = buffers[idx]
+        if buffer.count < capacity {
+            buffer.deallocate()
+            buffer = UnsafeMutableRawBufferPointer.allocate(byteCount: capacity, alignment: 0)
+            buffers[idx] = buffer
+        }
+        return buffer
+    }
+
+    mutating func reset() {
+        for idx in 0..<buffers.count {
+            buffers[idx].deallocate()
+        }
+    }
+}
+
+final class ChannelCompressionInboundHandler: ChannelInboundHandler {
```

Comment:
Maybe somehow reflect in the name of the class that it's about LZ4 compression?

Reply:
It is not supposed to use more than one compression type for now...
Can do it later if will change it.